nni/contrib/compression/pruning/basic_pruner.py

```diff
@@ -0,0 +1,194 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from __future__ import annotations
+
+import logging
+from typing import Dict, List, Literal, Tuple, overload
+
+import torch
+
+from .tools import _DATA, _METRICS, _MASKS, active_sparse_targets_filter, norm_metrics, fpgm_metrics, generate_sparsity
+from ..base.compressor import Compressor, Pruner
+from ..base.target_space import PruningTargetSpace
+from ..base.wrapper import ModuleWrapper
+from ..utils.evaluator import Evaluator
+
+_logger = logging.getLogger(__name__)
+
+
+class _NormPruner(Pruner):
+    p: int | str
+
+    @overload
+    def __init__(self, model: torch.nn.Module, config_list: List[Dict]):
+        ...
+
+    @overload
+    def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: Evaluator | None = None,
+                 existed_wrappers: Dict[str, ModuleWrapper] | None = None):
+        ...
+
+    def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: Evaluator | None = None,
+                 existed_wrappers: Dict[str, ModuleWrapper] | None = None):
+        super().__init__(model, config_list, evaluator, existed_wrappers)
+
+        # `skip_first_step`, `interval_steps` and `total_times` are used by `register_trigger`.
+        # `skip_first_step` controls if generating masks at the first step.
+        # `interval_steps` is the optimize step interval for generating masks.
+        # `total_times` is the total generation times of masks.
+        self.first_step_gen = False
+        self.interval_steps = -1
+        self.total_times: int | Literal['unlimited'] = 1
+
+    @classmethod
+    def from_compressor(cls, compressor: Compressor, new_config_list: List[Dict], evaluator: Evaluator | None = None):
+        return super().from_compressor(compressor, new_config_list, evaluator=evaluator)
+
+    def _collect_data(self) -> _DATA:
+        return active_sparse_targets_filter(self._target_spaces)
+
+    def _calculate_metrics(self, data: _DATA) -> _METRICS:
+        return norm_metrics(p=self.p, data=data, target_spaces=self._target_spaces)
+
+    def _generate_sparsity(self, metrics: _METRICS) -> _MASKS:
+        return generate_sparsity(metrics, self._target_spaces)
+
+    def _register_trigger(self, evaluator: Evaluator):
+        self._current_step = 0
+        self._is_first_step = True
+        self._remaining_times = self.total_times
+
+        def optimizer_task():
+            self._current_step += 1
+            if (self._is_first_step and self.first_step_gen) or self._current_step == self.interval_steps:
+                masks = self.generate_masks()
```

Comment:
where will the param first_step_gen will be updated？if the param firse_step_gen is False during the whole process, is it necessary to define it?

Reply:
nice catch, consider an iterative pruning process, norm pruning usually apply mask on the first step, here is a reserved interface for potential iterative pruning needs. No one uses this interface for now, I will add a comment for this attribute.